### PR TITLE
Content manager - Use condition_variable in retry wait loop

### LINKED ISSUE #20987: Improve content_manager wait loop
## Description 

It was found that the `content_manager` module could potentially prevent **wazuh-modulesd** from stopping because of the sleep method inside a loop. The thread can't read the `shouldRun` variable during the sleep

`src/shared_modules/content_manager/src/components/CtiDownloader.hpp`

```c++
... 
  while (retry && m_spUpdaterContext->spUpdaterBaseContext->shouldRun.load())
        {
            try
            {
                m_urlRequest.get(HttpURL(URL + queryParameters), onSuccess, onError, outputFilepath);
                retry = false;
            }
            catch (const cti_server_error& e)
            {
                constexpr auto SLEEP_TIME_THRESHOLD {30};

                logError(WM_CONTENTUPDATER, e.what());

                // Sleep and, if necessary, increase sleep time exponentially.
                std::this_thread::sleep_for(std::chrono::seconds(sleepTime));
                if (sleepTime < SLEEP_TIME_THRESHOLD)
                {
                    sleepTime = std::min(SLEEP_TIME_THRESHOLD, static_cast<int>(std::pow(2, retryAttempt)));
                    ++retryAttempt;
                }
            }
        }
```

A condition variable patter is preferred, see for example
https://github.com/wazuh/wazuh/pull/20967/files#diff-cf710bc1ea18828e323f6b95770aeb9ec6c4b394f619382e92a4090956a4808f

```c++
...
} while (!m_cv.wait_for(lock, std::chrono::seconds(delay), [&]() { return m_shouldStop.load(); }));
``` 
 
## DoD

- [x] Confirm the situation analyzed
- [x] Modify the loop to include the proposed pattern
- [x] Add UT 
- [ ] CI Pass

### PR DESCRIPTION
**Pull request by jftuduri**
|Related issue|
|---|
| Closes #20987|

## Description

This PR adds the class `ConditionSync`, a wrapper around a `std::condition_variable` that can be used to synchronize threads according to a boolean value.
This class is used in the Action Orchestration to signal when the action must be stopped. Using a condition_variable allows the different steps of the orchestration to respond immediately to the stop condition.

## Results

A fake server that always responds with a 505 error was used to force the retry mechanism.
<details><summary>Server script</summary>

```
from flask import Flask, request, jsonify

app = Flask(__name__)

@app.route('/endpoint', methods=['GET'])
def emulate_server():
    return jsonify({'error': 'Server Error'}), 505

if __name__ == '__main__':
    app.run(debug=True)
```

</details>

In the image, the retry mechanism was activated, prompting the content updater to attempt the download every 30 seconds. However, at 17:46:42, a stop request was issued, leading the content updater to promptly cancel its task and exit gracefully within a second, bypassing the usual 30-second waiting interval.
image

### CODE REVIEW

**Review by Damian-Mangold — COMMENTED** [review]
Partial review

**Review by tsarquis88 — COMMENTED** [review]
GJ!

---

Test: I've created a thread that waits 65 seconds and activates the stop flag, both in this branch and in dev-14153-vulndet-refactor branch:

- dev-14153-vulndet-refactor
```
[17:33:27] wazuh-modulesd:content-updater:[INFO]: Action for 'test' started
[17:33:27] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:actionOrchestrator.hpp:82 run: Running 'test' content update
[17:33:27] wazuh-modulesd:content-updater:[DEBUG]:CtiDownloader.hpp:203 handleRequest: CtiSnapshotDownloader - Starting process
[17:33:27] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:33:28] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:33:30] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:33:32] wazuh-modulesd:content-updater:[INFO]: Starting on-demand action for 'test'
[17:33:32] wazuh-modulesd:content-updater:[INFO]: Action in progress for 'test', on-demand request ignored
[17:33:32] 
[17:33:34] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:33:42] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:33:58] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:34:28] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:34:32] Killing process
[17:34:32] wazuh-modulesd:content-updater:[DEBUG]:onDemandManager.cpp:93 stopServer: Server stopped
[17:34:58] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:CtiDownloader.hpp:104 getCtiBaseParameters: CTI last offset: '0'
[17:34:58] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:CtiDownloader.hpp:105 getCtiBaseParameters: CTI last snapshot link: ''
[17:34:58] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:CtiDownloader.hpp:106 getCtiBaseParameters: CTI snapshot last offset: '0'
[17:34:58] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:CtiSnapshotDownloader.hpp:54 download: Downloading snapshot from ''
[17:34:58] wazuh-modulesd:content-updater:[DEBUG]:zipDecompressor.hpp:74 handleRequest: ZipDecompressor - Starting process
[17:34:58] wazuh-modulesd:content-updater:[DEBUG]:pubSubPublisher.hpp:61 handleRequest: PubSubPublisher - Starting process
[17:34:58] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:pubSubPublisher.hpp:49 publish: No data to publish
[17:34:58] wazuh-modulesd:content-updater:[DEBUG]:skipStep.hpp:48 handleRequest: SkipStep - Starting process
[17:34:58] wazuh-modulesd:content-updater:[DEBUG]:cleanUpContent.hpp:63 handleRequest: CleanUpContent - Starting process
[17:34:58] wazuh-modulesd:content-updater:[INFO]: Action for 'test' finished
[17:34:58] wazuh-modulesd:content-updater:[INFO]: Scheduler stopped for 'test'
```
> We can see that the orchestration was interrupted 28 seconds after the `Killing process` log.

- This branch
```
[17:40:14] wazuh-modulesd:content-updater:[INFO]: Action for 'test' started
[17:40:14] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:actionOrchestrator.hpp:82 run: Running 'test' content update
[17:40:14] wazuh-modulesd:content-updater:[DEBUG]:CtiDownloader.hpp:204 handleRequest: CtiSnapshotDownloader - Starting process
[17:40:14] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:40:15] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:40:17] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:40:19] wazuh-modulesd:content-updater:[INFO]: Starting on-demand action for 'test'
[17:40:19] wazuh-modulesd:content-updater:[INFO]: Action in progress for 'test', on-demand request ignored
[17:40:19] 
[17:40:21] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:40:29] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:40:45] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:41:15] wazuh-modulesd:content-updater:[ERROR]: Error 500 from server: HTTP response code said error
[17:41:19] Killing process
[17:41:19] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:CtiDownloader.hpp:104 getCtiBaseParameters: CTI last offset: '0'
[17:41:19] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:CtiDownloader.hpp:105 getCtiBaseParameters: CTI last snapshot link: ''
[17:41:19] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:CtiDownloader.hpp:106 getCtiBaseParameters: CTI snapshot last offset: '0'
[17:41:19] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:CtiSnapshotDownloader.hpp:54 download: Downloading snapshot from ''
[17:41:19] wazuh-modulesd:content-updater:[DEBUG]:zipDecompressor.hpp:74 handleRequest: ZipDecompressor - Starting process
[17:41:19] wazuh-modulesd:content-updater:[DEBUG]:onDemandManager.cpp:93 stopServer: Server stopped
[17:41:19] wazuh-modulesd:content-updater:[DEBUG]:pubSubPublisher.hpp:61 handleRequest: PubSubPublisher - Starting process
[17:41:19] wazuh-modulesd:content-updater:[DEBUG_VERBOSE]:pubSubPublisher.hpp:49 publish: No data to publish
[17:41:19] wazuh-modulesd:content-updater:[DEBUG]:skipStep.hpp:48 handleRequest: SkipStep - Starting process
[17:41:19] wazuh-modulesd:content-updater:[DEBUG]:cleanUpContent.hpp:63 handleRequest: CleanUpContent - Starting process
[17:41:19] wazuh-modulesd:content-updater:[INFO]: Action for 'test' finished
[17:41:19] wazuh-modulesd:content-updater:[INFO]: Scheduler stopped for 'test'
```
> We can see that the orchestration was interrupted right after the `Killing process` log.

File: src/shared_modules/content_manager/src/components/conditionSync.hpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        return m_condition.load();
+    }
+
+    /**
+     * @brief Sets the value of the condition to the specified value.
+     * @param value The new value of the condition.
+     *
+     * Sets the value of the condition to the specified value and notifies
+     * all waiting threads.
```
[the remark this answers is not in the document]
Comment (jftuduri) [reply src/shared_modules/content_manager/src/components/conditionSync.hpp]: Kind of. Fixed!

**Review by tsarquis88 — APPROVED** [review]
LGTM! 

Good idea with the new class, btw.

**Review by pereyra-m — APPROVED** [review]
LGTM!

**Review by Damian-Mangold — APPROVED** [review]
LGTM !

**Pull request merged**